File: go/pkg/trees/trees_1.go
package trees

import (
	"fmt"
	"strconv"
)

// TreeNode represents a binary tree node
type TreeNode struct {
	Data  int
	Left  *TreeNode
	Right *TreeNode
}

// ToString provides a string representation for binary tree node
func (n *TreeNode) ToString() string {
	if n == nil {
		return "nil"
	}
	if n.Left == nil && n.Right == nil {
		return strconv.Itoa(n.Data)
	}
	return fmt.Sprintf("%d => (%s, %s)", n.Data, n.Left.ToString(), n.Right.ToString())
}

// O(1)
func splitSlice(ns []int) (middle int, left []int, right []int) {
	i := len(ns) / 2
	return ns[i], ns[:i], ns[i+1:]
}

// MinBST runs in O(n)
func MinBST(ns []int, node *TreeNode) *TreeNode {
	if len(ns) == 0 {
		return nil
	}

	if node == nil {
		node = new(TreeNode)
	}

	data, left, right := splitSlice(ns)
	node.Data = data

	if len(left) > 0 {
		node.Left = new(TreeNode)
		MinBST(left, node.Left)
	}

	if len(right) > 0 {
		node.Right = new(TreeNode)
		MinBST(right, node.Right)
	}

	return node
}

File: python/trees/trees_1.py
class TreeNode:
    def __init__(self, data=None, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    def __repr__(self):
        if not self.left and not self.right:
            return str(self.data)
        return f'{self.data} => ({self.left}, {self.right})'


# O(1) return the middle of the list and a reference
# to left and right sublists not including the middle item
def splice_middle(l):
    i = len(l) // 2
    return l[i], l[:i], l[i+1:]


# O(n) one recursive call per item in the list
def min_bst(l, node=TreeNode()):
    if len(l) == 0:
        return

    data, left, right = splice_middle(l)
    node.data = data

    if len(left):
        node.left = TreeNode()
        min_bst(left, node.left)
    if len(right):
        node.right = TreeNode()
        min_bst(right, node.right)

    return node
